README.md
[![Build Status](https://travis-ci.org/orjail/orjail.svg?branch=master)](https://travis-ci.org/orjail/orjail)

> ### :warning: WARNING
> Security isn’t just about the tools you use or the software you download. It begins with understanding the unique threats you face and how you can counter those threats.

# orjail
**orjail** is a tool that lets you create a jail around a program to force its network traffic through [Tor](https://www.torproject.org/). 
It creates a hostile environment for anything trying to discover your real IP address.


## Install
##### From source
```
git clone https://github.com/orjail/orjail.git
cd orjail
sudo make install
```
##### Debian / Ubuntu
We provide a `.deb` package you can download from
[here](https://github.com/orjail/orjail/releases/latest) and install with a
simple:


```
sudo dpkg -i orjail*.deb
```

## Why?
We've tried to de-anonymize a program executed in a torsocks environment and that was not so difficult as torsocks uses LD_PRELOAD, so you only need to statically compile your stuff.
As [Whonix](https://www.whonix.org/) is sometimes too much, the idea is to experiment with [linux namespaces](http://man7.org/linux/man-pages/man7/namespaces.7.html) and learn by doing something useful (at least for us).

## Requirements
- Linux kernel supporting namespaces (you have it since 2008)
- Tor installed
- [firejail](https://firejail.wordpress.com/) (optional, but really suggested)

## How it works
It creates a separated [network namespace](https://en.wikipedia.org/wiki/Linux_namespaces#Network_(net)) (using `ip netns`) with its own network
interface and a link to the host interface with some iptables rules (on host)
that forces traffic generated from inside orjail to only exit via Tor (including DNS).  
Inside **orjail** you'll be in another pid namespace (try `sudo orjail ps aux`) and another mount namespace (we use this to show a different /etc/resolv.conf).  

**if you find a way to de-anonymize a program running inside orjail** (also a shell with root privileges) it would be nice to [share it with us](https://github.com/orjail/orjail/issues)


## Additional info
- orjail needs root permissions to run
- orjail runs your command as your user
- orjail will launch a Tor instance bound to orjail interface

## Usage
> ```bash
> orjail [options] [command]
> ```
> **-u, --user** \<user>  
> Run command as \<user> (default **$USER**)
>
> **-f, --firejail**  
> Use [firejail](https://firejail.wordpress.com) as a security container
>
> **--firejail-args** "\<args>"  
> Set arguments to pass to firejail surrounded by quotes.  
> eg. "--hostname=host --env=PS1=[orjail]"
>
>
> **--host-torrc**  
> Include your torrc host
>
> **-t, --tor-exec** \<torpath>  
> Select a Tor executable to use. The path can be full, relative or be in $PATH (default **tor**)
>
> **-s, --shell**  
> Execute a shell (default **$SHELL**)
>
>
> **-k, --keep**  
> Don't delete namespace and don't kill tor after the execution.
>
> **-n, --name <name>**  
> Set a custom namespace name (default **orjail**)

## Example

##### An example to understand what are we talking about:
```bash
$ sudo orjail ifconfig
out-orjail: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.200.1.2  netmask 255.255.255.0  broadcast 0.0.0.0
        inet6 fe80::6439:afff:febc:c9b5  prefixlen 64  scopeid 0x20<link>
        ether 66:39:af:bc:c9:b5  txqueuelen 1000  (Ethernet)
        RX packets 6847  bytes 7488116 (7.1 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 6809  bytes 915088 (893.6 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

#### Get homepage content with curl via Tor
`sudo orjail curl autistici.org > autistici.org `

#### Same as before with another user
`sudo orjail -u another_user curl autistici.org`

#### "Resolve" an onion address (not so useful, just to show that .onion resolving works)
`sudo orjail dig wi7qkxyrdpu5cmvr.onion`

#### Run a hidden service inside orjail (you'll find your address inside `examples/hostname`)
`sudo orjail -v -H 8080 -d examples  "python -m SimpleHTTPServer 8080" `

#### Get an onion webserver content via Tor:
`sudo orjail curl wi7qkxyrdpu5cmvr.onion`

#### Open a firefox that could reach internet via Tor only:
`sudo orjail firefox -P /tmp/tmpprofile`

> ### :warning:
> firefox has a flag that blocks .onion resolution by default, change it in **about:config**/**network.dns.blockDotOnion**.  
>
> **running a browser inside orjail is not safe, please use [Tor Browser](https://www.torproject.org/projects/torbrowser.html.en) instead**

#### Get an anonymous shell
`sudo orjail -s`

#### Run pidgin in verbose mode
`sudo orjail -v pidgin`

#### Keep the namespace after exit so we can start another program in same ns 
`sudo orjail -k ls`

#### Use `firejail` as a security sandbox to join orjail network namespace
`sudo orjail -f thunderbird`

## Known issues
- dbus
- X


---
Made with  :heart: by [_to hacklab](https://autistici.org/underscore)


debian/changelog
orjail (1:1.1-1) unstable; urgency=medium

  * New upstream version (local package).

 -- lesion <lesion@autistici.org>  Tue, 12 Feb 2019 15:21:35 +0000

orjail (1:1.0-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@riseup.net>  Sun, 19 Aug 2018 14:30:00 +0000

orjail (1:0.9-1) unstable; urgency=low

  * Initial release.

 -- Patrick Schleizer <adrelanos@riseup.net>  Sun, 19 Aug 2018 14:00:00 +0000


netjail.sh
#!/bin/bash

NAME=netjail

## remove/add tor network namespace :)
#sudo ip netns del tor

#
# check if network namespace already exists
sudo ip netns list | grep -e ^$NAME\  &> /dev/null
if [ $? -eq 0 ]; then
  echo "$NAME network namespace already exists"
else
  # add network namespace
  sudo ip netns add $NAME

  # Create veth link.
  sudo ip link add in-$NAME type veth peer name out-$NAME

  # Add out to NS.
  sudo ip link set out-$NAME netns $NAME

  ## set ip address
  sudo ip addr add 10.200.1.1/24 dev in-$NAME
  sudo ip link set in-$NAME up

  # Setup IP address of v-peer1.
  sudo ip netns exec $NAME ip addr add 10.200.1.2/24 dev out-$NAME
  sudo ip netns exec $NAME ip link set out-$NAME up

  # default route
  sudo ip netns exec $NAME ip route add default via 10.200.1.1

  # resolve with tor
  sudo iptables -t nat -A  PREROUTING -s 10.200.1.0/24 -p udp --dport 53 -j REDIRECT --to-ports 5353

  # traffic througth tor
  sudo iptables -t nat -A PREROUTING -s 10.200.1.0/24 -p tcp --syn  -j REDIRECT --to-ports 9040    
  sudo iptables -A OUTPUT -m state --state ESTABLISHED,RELATED -j ACCEPT

fi

ip netns exec $NAME $*

#sudo iptables -A OUTPUT -m state --state ESTABLISHED,RELATED -j ACCEPT


torjail.sh
#!/bin/bash

REALPATH=`realpath $0`
REALPATH=`dirname $REALPATH`
DEFAULTNAME=torjail
NAME=$DEFAULTNAME
USERNAME=`whoami`

# Functions
# ~~~~~~~~~

print_real() {
  if [ $2 = 'G' ]; then
    echo $1 -e "\e[32m$3\e[0m"
  elif [ $2 = 'Y' ]; then
    echo $1 -e "\e[33m$3\e[0m"
  elif [ $2 = 'N' ]; then
    echo $1 "$3"
  else
    echo $1 -e "\e[31m$3\e[0m"
  fi
}

print() {
  print_real '' "$1" "$2"
}

printn() {
  print_real "-n" "$1" "$2"
}

help_and_exit() {
  printn N "Usage: "
  printn Y "$0"
  print G " <options> [command <arguments>...]"
  print N "Options:"
  print N "    -h, --help         It shows this menu."
  print N "    -n, --name <name>  Set a custom namespace name. By default '$DEFAULTNAME'."
  print N "If command is not passed, the shell will be executed."
  exit 0
}

# Inside part
# ~~~~~~~~~~~

# This part is executed only inside the namespace.
if [ "$1" = "--inside" ]; then
  REALPATH=`realpath $0`
  REALPATH=`dirname $REALPATH`

  shift
  USERNAME="$1"
  shift

  print G " * Replacing resolv.conf..."
  mount --bind $REALPATH/resolv.conf /etc/resolv.conf

  print G " * Executing..."

  sudo -u $USERNAME $*
  exit
fi

# The tool
# ~~~~~~~~

# Arguments check
while [[ $# -gt 0 ]]; do
  key="$1"
  case $key in
    # Replacing the name
    -n|--name)
      NAME="$2"
      if [ "$NAME" = "" ]; then
        help_and_exit
      fi
      shift
    ;;

    # Help menu
    -h|--help)
      help_and_exit
    ;;

    # The rest
    *)
    break
    ;;
  esac
  shift
done

# Let's replace $1 with the current shell if no command has been passed.
if [ "$1" = "" ]; then
  set -- "$SHELL"
fi

# check if network namespace already exists
sudo ip netns list | grep -e ^$NAME\  &> /dev/null
if [ $? -ne 0 ]; then
  print G "It seems that you don't have the namespace $NAME."
  printn Y "Do you want to create it? [y/n] "
  read CREATE

  if [ "$CREATE" != y ] && [ "$CREATE" != Y ]; then
    print G "Ok. Bye!"
    exit 0
  fi

  print G " * Creating a $NAME namespace..."
  # add network namespace
  sudo ip netns add $NAME

  # Create veth link.
  print G " * Creating a veth link..."
  sudo ip link add in-$NAME type veth peer name out-$NAME

  # Add out to NS.
  print G " * Sharing the veth interface..."
  sudo ip link set out-$NAME netns $NAME

  ## setup ip address of host interface
  print G " * Setting up IP address of host interface..."
  sudo ip addr add 10.200.1.1/24 dev in-$NAME
  sudo ip link set in-$NAME up

  # setup ip address of peer
  print G " * Setting up IP address of peer interface..."
  sudo ip netns exec $NAME ip addr add 10.200.1.2/24 dev out-$NAME
  sudo ip netns exec $NAME ip link set out-$NAME up

  # default route
  print G " * Default routing up..."
  sudo ip netns exec $NAME ip route add default via 10.200.1.1

  # resolve with tor
  print G " * Resolving via TOR..."
  sudo iptables -t nat -A  PREROUTING -i in-$NAME -p udp --dport 53 -j REDIRECT --to-ports 5353

  # traffic througth tor
  print G " * Traffic via TOR..."
  sudo iptables -t nat -A PREROUTING -i in-$NAME -p tcp --syn  -j REDIRECT --to-ports 9040
  sudo iptables -A OUTPUT -m state --state ESTABLISHED,RELATED -j ACCEPT

  print G "All done!"
else
  print Y "$NAME network namespace already exists!"
fi

# run your shit
sudo ip netns exec $NAME unshare --ipc --fork --pid --mount --mount-proc $0 --inside $USERNAME $*
